src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -1475,1524 +1574,1682 @@ int16_t LoRaWANNode::downlink(uint8_t* data, size_t* len, LoRaWANEvent_t* event)
+  }
+  downlinkMsg[RADIOLIB_LORAWAN_BLOCK_DIR_POS] = RADIOLIB_LORAWAN_DOWNLINK;
+  LoRaWANNode::hton<uint32_t>(&downlinkMsg[RADIOLIB_LORAWAN_BLOCK_DEV_ADDR_POS], this->devAddr);
+  LoRaWANNode::hton<uint16_t>(&downlinkMsg[RADIOLIB_LORAWAN_BLOCK_FCNT_POS], fCnt32);
+  downlinkMsg[RADIOLIB_LORAWAN_MIC_BLOCK_LEN_POS] = downlinkMsgLen - sizeof(uint32_t);
+
+  // check the MIC
+  if(!verifyMIC(downlinkMsg, RADIOLIB_AES128_BLOCK_SIZE + downlinkMsgLen, this->sNwkSIntKey)) {
+    #if !RADIOLIB_STATIC_ONLY
+      delete[] downlinkMsg;
+    #endif
+    return(RADIOLIB_ERR_CRC_MISMATCH);
+  }
+  
+  // save current fCnt to respective frame counter
+  if (isAppDownlink) {
+    this->aFCntDown = fCnt32;
+  } else {
+    this->nFCntDown = fCnt32;
+  }
+
+  RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Downlink (%sFCntDown = %lu) encoded:", 
+                                  isAppDownlink ? "A" : "N", 
+                                  (unsigned long)(isAppDownlink ? this->aFCntDown : this->nFCntDown));
+  RADIOLIB_DEBUG_PROTOCOL_HEXDUMP(downlinkMsg, RADIOLIB_AES128_BLOCK_SIZE + downlinkMsgLen);
+
+  // if this is a confirmed frame, save the downlink number (only app frames can be confirmed)
+  bool isConfirmedDown = false;
   if((downlinkMsg[RADIOLIB_LORAWAN_FHDR_LEN_START_OFFS] & 0xFE) == RADIOLIB_LORAWAN_MHDR_MTYPE_CONF_DATA_DOWN) {
     this->confFCntDown = this->aFCntDown;
     isConfirmedDown = true;
   }
+
+  // a downlink was received, so reset the ADR counter to the last uplink's fCnt
+  this->adrFCnt = this->getFCntUp();
   
-  // pass the extra info if requested
-  if(event) {
-    event->dir = RADIOLIB_LORAWAN_CHANNEL_DIR_DOWNLINK;
-    event->confirmed = isConfirmedDown;
-    event->confirming = isConfirmingUp;
-    event->datarate = this->dataRates[RADIOLIB_LORAWAN_CHANNEL_DIR_DOWNLINK];
-    event->freq = currentChannels[event->dir].freq;
-    event->power = this->txPowerMax - this->txPowerSteps * 2;
-    event->fCnt = isAppDownlink ? this->aFCntDown : this->nFCntDown;
-    event->fPort = fPort;
+  // if this downlink is on FPort 0, the FOptsLen is the length of the payload
+  // in any other case, the payload (length) is user accessible
+  uint8_t fOptsLen = fOptsPbLen;
+  if(fPort == RADIOLIB_LORAWAN_FPORT_MAC_COMMAND && payLen > 0) {
+    fOptsLen = payLen;
+  } else {
+    *len = payLen;
   }
 
-  // clear the previous MAC commands, if any
-  memset(&(this->commandsDown), 0, sizeof(LoRaWANMacCommandQueue_t));
-  
-  // process FOpts (if there are any)
-  if(fOptsLen > 0) {
-    // there are some Fopts, decrypt them
-    #if !RADIOLIB_STATIC_ONLY
-      uint8_t* fOpts = new uint8_t[RADIOLIB_MAX(RADIOLIB_LORAWAN_FHDR_FOPTS_LEN_MASK, (int)fOptsLen)];
-    #else
-      uint8_t fOpts[RADIOLIB_STATIC_ARRAY_SIZE];
-    #endif
+  #if !RADIOLIB_STATIC_ONLY
+    uint8_t* fOpts = new uint8_t[fOptsLen];
+  #else
+    uint8_t fOpts[RADIOLIB_STATIC_ARRAY_SIZE];
+  #endif
```

Comment:
I seem to recall there can be no more than 16 fOpts bytes in LW 1.1. If that is also the case for 1.0.x, it would make more sense to just have 16B buffer here instead of dynamic memory.

Reply:
That is true for piggybacked MAC commands (in the FOpts field of the LoRaWAN header), but if FPort = 0, the whole payload consists of FOpts with theoretically up to 230-ish bytes.

Reply:
Ah, I see - thanks!